### PR TITLE
Fix races in CanAccessDirectory

### PR DESCRIPTION
**Pull request by nikola-sh**
If several bazel processes are starting concurrently some of them can fail to start with error
```Output base directory ... must be readable and writable.```

It happens because CanAccessDirectory doesn't handle `NTSTATUS` `STATUS_DELETE_PENDING` for file.
`STATUS_DELETE_PENDING` occured if `DeleteFile` finished successfully but there are open handles for the file.
`STATUS_DELETE_PENDING` is converted to `ERROR_ACCESS_DENIED` on Win32 level so client code can't distinguish pending deletion from real denied access.

To workaround this problem we can create check-file with unique name, for example with current thread id.

P.S.
This problem is reproducible on Windows versions prior to Windows 10 19H1.

### CODE REVIEW

File: src/main/cpp/util/file_windows.cc [hunk 1]
```diff
@@ -548,9 +552,9 @@ bool CanAccessDirectory(const Path& path) {
       /* dwFlagsAndAttributes */ FILE_ATTRIBUTE_NORMAL,
       /* hTemplateFile */ NULL);
   DWORD err = GetLastError();
-  if (handle == INVALID_HANDLE_VALUE && err != ERROR_ALREADY_EXISTS) {
```
Comment (nikola-sh) [inline src/main/cpp/util/file_windows.cc]: We should respect any error here, `handle == INVALID_HANDLE_VALUE && err==ERROR_ALREADY_EXISTS` is incorrect invariant for `OPEN_ALWAYS`.

Comment (laszlocsomor) [reply src/main/cpp/util/file_windows.cc]: Why incorrect? CreateFileW's docs says: "If the specified file exists, the function succeeds and the last-error code is set to ERROR_ALREADY_EXISTS (183)."

Comment (nikola-sh) [reply src/main/cpp/util/file_windows.cc]: Yes, but if function succeeds handle != INVALID_HANDLE_VALUE

Comment (laszlocsomor) [reply src/main/cpp/util/file_windows.cc]: Ah, of course.

File: src/main/cpp/util/file_windows.cc [hunk 2]
```diff
@@ -534,7 +535,10 @@ bool CanAccessDirectory(const Path& path) {
 
   // The only easy way to know if a directory is writable is by attempting to
   // open a file for writing in it.
-  Path dummy_path = path.GetRelative("bazel_directory_access_test");
+  std::string dummy_name = "bazel_directory_access_test_";
+  dummy_name += std::to_string(::GetCurrentThreadId());
```
Comment (laszlocsomor) [inline src/main/cpp/util/file_windows.cc]: Please add a brief comment here saying that the Thread ID avoids races among concurrent Bazel processes.

Comment (nikola-sh) [reply src/main/cpp/util/file_windows.cc]: @laszlocsomor can you restart checks? They are failed on platforms that I didn't change.

**Comment by laszlocsomor** [comment]
Don't worry, I know the tests aren't your PR's fault.
I already imported the PR, just waiting for internal review.

**Pull request closed without merging**